lib/Sema/CSApply.cpp

```diff
@@ -4715,15 +4715,21 @@ namespace {
       // Resolve each of the components.
       bool didOptionalChain = false;
       bool isFunctionType = false;
-      Type baseTy, leafTy;
+      auto baseTy = cs.hasRootType(E)
+                        ? cs.simplifyType(cs.getRootType(E))
+                        : Type();
+      Type leafTy;
       Type exprType = cs.getType(E);
+      // Extract the root and value type from the inferred type of the
+      // expression. (If we already have an explicit root type stored in the
+      // constraint system, use that as the source of truth.)
```

Comment:
CSGen records a root type for every KeyPathExpr it finds, in what situations `baseTy` is going to be empty?

Reply:
Ah yeah good call. I think this was left from an earlier strategy that didn't _always_ record a type but I think you're right it's unneeded now.